File: scripts/semantic_eval.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.10"
# dependencies = []
# ///
"""Offline semantic eval command for captured skill fixtures.

This runner evaluates fixture text and seeded artifact facts only. It never
invokes Claude Code, OpenCode, or any other model runtime.
"""

from __future__ import annotations

import argparse
import json
import sys
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from semantic_fixtures import SemanticFixture, load_fixture


@dataclass(frozen=True)
class CheckedFact:
    """One deterministic semantic assertion result."""

    fact: str
    status: str
    detail: str

    def as_dict(self) -> dict[str, str]:
        return {"fact": self.fact, "status": self.status, "detail": self.detail}


def evaluate_fixture(fixture: SemanticFixture, source: str = "<fixture>") -> dict[str, Any]:
    """Evaluate one validated fixture and return a machine-readable result."""
    facts = _check_output_facts(fixture) + _check_seeded_artifact_facts(fixture)
    failing = next((fact for fact in facts if fact.status == "fail"), None)
    return {
        "fixture": source,
        "status": "fail" if failing else "pass",
        "checked_facts": [fact.as_dict() for fact in facts],
        "failing_fact": failing.as_dict() if failing else None,
    }


def evaluate_fixture_file(path: Path) -> dict[str, Any]:
    """Load, validate, and evaluate a fixture file."""
    fixture, errors = load_fixture(path)
    if errors:
        failing = CheckedFact("fixture_contract", "fail", "; ".join(errors))
        return {
            "fixture": str(path),
            "status": "fail",
            "checked_facts": [failing.as_dict()],
            "failing_fact": failing.as_dict(),
        }
    assert fixture is not None
    return evaluate_fixture(fixture, str(path))


def build_report(results: list[dict[str, Any]]) -> dict[str, Any]:
    """Build the command's stable JSON summary."""
    passed = sum(1 for result in results if result["status"] == "pass")
    failed = len(results) - passed
    return {
        "timestamp": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "status": "fail" if failed else "pass",
        "fixtures_tested": len(results),
        "passed": passed,
        "failed": failed,
        "results": results,
    }


def _check_output_facts(fixture: SemanticFixture) -> list[CheckedFact]:
    facts: list[CheckedFact] = []
    expected = fixture.expected_facts
    for index, text in enumerate(expected.get("required_output", [])):
        found = text in fixture.captured_output
        facts.append(CheckedFact(
            f"required_output[{index}]",
            "pass" if found else "fail",
            f"captured output {'contains' if found else 'does not contain'} {text!r}",
        ))
    for index, text in enumerate(expected.get("forbidden_output", [])):
        found = text in fixture.captured_output
        facts.append(CheckedFact(
            f"forbidden_output[{index}]",
            "fail" if found else "pass",
            f"captured output {'contains forbidden' if found else 'omits forbidden'} {text!r}",
        ))

    writes = expected.get("artifact_expectations", {}).get("writes")
    if writes == "none":
        facts.append(CheckedFact(
            "artifact_expectations.writes",
            "pass",
            "fixture expects no artifact writes; offline eval performed none",
        ))
    return facts


def _check_seeded_artifact_facts(fixture: SemanticFixture) -> list[CheckedFact]:
    facts: list[CheckedFact] = []
    by_path = {
        item["path"]: item["content"]
        for item in fixture.seeded_state.get("files", [])
        if isinstance(item, dict) and isinstance(item.get("path"), str)
    }

    for index, expected in enumerate(fixture.expected_facts.get("required_artifacts", [])):
        fact_name = f"required_artifacts[{index}]"
        if not isinstance(expected, dict):
            facts.append(CheckedFact(fact_name, "fail", "expected artifact fact must be an object"))
            continue

        path = expected.get("path")
        if not isinstance(path, str) or not path.strip():
            facts.append(CheckedFact(fact_name, "fail", "expected artifact fact must name a path"))
            continue

        content = by_path.get(path)
        if content is None:
            facts.append(CheckedFact(fact_name, "fail", f"seeded artifact {path!r} is missing"))
            continue

        missing = [text for text in expected.get("contains", []) if text not in content]
        if missing:
            facts.append(CheckedFact(fact_name, "fail", f"seeded artifact {path!r} lacks {missing[0]!r}"))
        else:
            facts.append(CheckedFact(fact_name, "pass", f"seeded artifact {path!r} matched"))
    return facts


def parse_args(argv: list[str] | None = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Offline semantic eval runner for captured Agentera skill fixtures."
    )
    parser.add_argument("fixtures", nargs="+", type=Path, help="Semantic fixture Markdown files to evaluate.")
    return parser.parse_args(argv)


def main(argv: list[str] | None = None) -> int:
    args = parse_args(argv)
    report = build_report([evaluate_fixture_file(path) for path in args.fixtures])
    print(json.dumps(report, indent=2))
    return 0 if report["status"] == "pass" else 1


if __name__ == "__main__":
    sys.exit(main())

File: packages/cli/src/eval/semanticEval.ts
import { SemanticFixture, loadFixture, pyRepr } from "./semanticFixtures.js";
import { pyJsonIndent } from "../core/pyjson.js";
import type { JsonObject } from "../core/jsonValue.js";

/**
 * Offline semantic eval runner for captured skill fixtures. Faithful TS port of
 * scripts/semantic_eval.py. Never invokes a model runtime.
 */

export interface CheckedFact {
  fact: string;
  status: string;
  detail: string;
}

function utcTimestamp(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

export function evaluateFixture(fixture: SemanticFixture, source = "<fixture>"): JsonObject {
  const facts = [
    ...checkOutputFacts(fixture),
    ...checkToolTraceFacts(fixture),
    ...checkSeededArtifactFacts(fixture),
  ];
  const failing = facts.find((fact) => fact.status === "fail") ?? null;
  return {
    fixture: source,
    status: failing ? "fail" : "pass",
    checked_facts: facts.map((f) => ({ fact: f.fact, status: f.status, detail: f.detail })),
    failing_fact: failing ? { fact: failing.fact, status: failing.status, detail: failing.detail } : null,
  };
}

export function evaluateFixtureFile(path: string): JsonObject {
  const [fixture, errors] = loadFixture(path);
  if (errors.length > 0) {
    const failing = { fact: "fixture_contract", status: "fail", detail: errors.join("; ") };
    return {
      fixture: path,
      status: "fail",
      checked_facts: [failing],
      failing_fact: failing,
    };
  }
  return evaluateFixture(fixture!, path);
}

export function buildReport(results: JsonObject[]): JsonObject {
  const passed = results.filter((result) => result.status === "pass").length;
  const failed = results.length - passed;
  return {
    timestamp: utcTimestamp(),
    status: failed ? "fail" : "pass",
    fixtures_tested: results.length,
    passed,
    failed,
    results,
  };
}

function checkOutputFacts(fixture: SemanticFixture): CheckedFact[] {
  const facts: CheckedFact[] = [];
  const expected = fixture.expectedFacts;
  const requiredOutput = expected.required_output;
  const requiredOutputArr: string[] = Array.isArray(requiredOutput) ? (requiredOutput as string[]) : []; // cast: JSON.parse fixture IO boundary
  requiredOutputArr.forEach((text: string, index: number) => {
    const found = fixture.capturedOutput.includes(text);
    facts.push({
      fact: `required_output[${index}]`,
      status: found ? "pass" : "fail",
      detail: `captured output ${found ? "contains" : "does not contain"} ${pyRepr(text)}`,
    });
  });
  const forbiddenOutput = expected.forbidden_output;
  const forbiddenOutputArr: string[] = Array.isArray(forbiddenOutput) ? (forbiddenOutput as string[]) : []; // cast: JSON.parse fixture IO boundary
  forbiddenOutputArr.forEach((text: string, index: number) => {
    const found = fixture.capturedOutput.includes(text);
    facts.push({
      fact: `forbidden_output[${index}]`,
      status: found ? "fail" : "pass",
      detail: `captured output ${found ? "contains forbidden" : "omits forbidden"} ${pyRepr(text)}`,
    });
  });

  const artifactExpectations = expected.artifact_expectations as JsonObject | null; // cast: JSON.parse fixture IO boundary
  const writes = artifactExpectations?.writes;
  if (writes === "none") {
    facts.push({
      fact: "artifact_expectations.writes",
      status: "pass",
      detail: "fixture expects no artifact writes; offline eval performed none",
    });
  }
  return facts;
}

function checkToolTraceFacts(fixture: SemanticFixture): CheckedFact[] {
  const facts: CheckedFact[] = [];
  const expected = fixture.expectedFacts;
  const rawCalls = fixture.toolTrace.calls;
  const callList: string[] = Array.isArray(rawCalls) ? (rawCalls as string[]) : []; // cast: JSON.parse fixture IO boundary
  const calls = callList.join("\n");
  const requiredToolCalls = expected.required_tool_calls;
  const requiredToolCallsArr: string[] = Array.isArray(requiredToolCalls) ? (requiredToolCalls as string[]) : []; // cast: JSON.parse fixture IO boundary
  requiredToolCallsArr.forEach((text: string, index: number) => {
    const found = calls.includes(text);
    facts.push({
      fact: `required_tool_calls[${index}]`,
      status: found ? "pass" : "fail",
      detail: `tool trace ${found ? "contains" : "does not contain"} ${pyRepr(text)}`,
    });
  });
  const forbiddenToolCalls = expected.forbidden_tool_calls;
  const forbiddenToolCallsArr: string[] = Array.isArray(forbiddenToolCalls) ? (forbiddenToolCalls as string[]) : []; // cast: JSON.parse fixture IO boundary
  forbiddenToolCallsArr.forEach((text: string, index: number) => {
    const found = calls.includes(text);
    facts.push({
      fact: `forbidden_tool_calls[${index}]`,
      status: found ? "fail" : "pass",
      detail: `tool trace ${found ? "contains forbidden" : "omits forbidden"} ${pyRepr(text)}`,
    });
  });
  const toolCallCounts = expected.tool_call_counts;
  if (toolCallCounts && typeof toolCallCounts === "object" && !Array.isArray(toolCallCounts)) {
    for (const [text, expectedCount] of Object.entries(toolCallCounts as JsonObject)) { // cast: JSON.parse fixture IO boundary
      const actual = callList.filter((call) => call.includes(text)).length;
      facts.push({
        fact: `tool_call_counts[${text}]`,
        status: actual === expectedCount ? "pass" : "fail",
        detail: `tool trace contains ${actual} call(s) matching ${pyRepr(text)}; expected ${expectedCount}`,
      });
    }
  }
  return facts;
}

function checkSeededArtifactFacts(fixture: SemanticFixture): CheckedFact[] {
  const facts: CheckedFact[] = [];
  const byPath: Record<string, string> = {};
  const rawFiles = fixture.seededState.files;
  const filesArr: unknown[] = Array.isArray(rawFiles) ? rawFiles : [];
  for (const item of filesArr) {
    if (item && typeof item === "object" && !Array.isArray(item)) {
      const obj = item as JsonObject; // cast: JSON.parse fixture IO boundary
      const path = obj.path;
      if (typeof path === "string") {
        byPath[path] = obj.content as string; // cast: JSON.parse fixture IO boundary
      }
    }
  }

  const requiredArtifacts = fixture.expectedFacts.required_artifacts;
  const requiredArtifactsArr: unknown[] = Array.isArray(requiredArtifacts) ? requiredArtifacts : [];
  requiredArtifactsArr.forEach((expected: unknown, index: number) => {
    const factName = `required_artifacts[${index}]`;
    if (!expected || typeof expected !== "object" || Array.isArray(expected)) {
      facts.push({ fact: factName, status: "fail", detail: "expected artifact fact must be an object" });
      return;
    }
    const expectedObj = expected as JsonObject; // cast: JSON.parse fixture IO boundary
    const path = expectedObj.path;
    if (typeof path !== "string" || !path.trim()) {
      facts.push({ fact: factName, status: "fail", detail: "expected artifact fact must name a path" });
      return;
    }
    const content = byPath[path];
    if (content === undefined) {
      facts.push({ fact: factName, status: "fail", detail: `seeded artifact ${pyRepr(path)} is missing` });
      return;
    }
    const containsRaw = expectedObj.contains;
    const containsArr: string[] = Array.isArray(containsRaw) ? (containsRaw as string[]) : []; // cast: JSON.parse fixture IO boundary
    const missing = containsArr.filter((text: string) => !content.includes(text));
    if (missing.length > 0) {
      facts.push({ fact: factName, status: "fail", detail: `seeded artifact ${pyRepr(path)} lacks ${pyRepr(missing[0])}` });
    } else {
      facts.push({ fact: factName, status: "pass", detail: `seeded artifact ${pyRepr(path)} matched` });
    }
  });
  return facts;
}

export function main(argv: string[] = [], out: (line: string) => void = (l) => process.stdout.write(l)): number {
  if (argv.length === 0) {
    process.stderr.write("usage: semantic_eval <fixtures...>\n");
    return 2;
  }
  const report = buildReport(argv.map((path) => evaluateFixtureFile(path)));
  out(pyJsonIndent(report) + "\n");
  return report.status === "pass" ? 0 : 1;
}
